### PR TITLE
Further refinement of {Float,Double,Float80}.init(_:String)

### PR DESCRIPTION
**Pull request by tbkka**
After talking with @stephentyrone, I found some additional simplifications. No
functional change, just shorter/simpler.

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Review by stephentyrone — APPROVED** [review]

File: stdlib/public/core/FloatingPointParsing.swift.gyb [hunk 1]
```diff
@@ -126,36 +126,23 @@ extension ${Self}: LosslessStringConvertible {
...
+      text.withCString { chars -> Bool in
+        switch chars[0] {
+        case 9, 10, 11, 12, 13, 32:
+          return false // Reject leading whitespace
+        case 0:
+          return false // Reject empty string
+        default:
+          break
+        }
+        let endPtr = _swift_stdlib_strto${cFuncSuffix2[bits]}_clocale(chars, p)
```
Comment (airspeedswift) [inline stdlib/public/core/FloatingPointParsing.swift.gyb]: `_swift_stdlib_strto...` has existed since 5.0, right?

Comment (airspeedswift) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: oh ignore me, it was there already

Comment (stephentyrone) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: Yes, no ABI change here.

File: stdlib/public/core/FloatingPointParsing.swift.gyb [hunk 2]
```diff
@@ -126,36 +126,23 @@ extension ${Self}: LosslessStringConvertible {
   ///   is `nil`.
   @inlinable // FIXME(sil-serialize-all)
```
Comment (airspeedswift) [inline stdlib/public/core/FloatingPointParsing.swift.gyb]: thinking of this comment (that we should probably remove...) could we use `@_specialize` and then make this non-inlinable to reduce having to inline it?

Comment (airspeedswift) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: given there are only 2 string protocol conformance rn

Comment (stephentyrone) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: Not unreasonable.

Comment (jckarter) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: Even better, make the generic thing a stub that just canonicalizes to `Substring` and make the non-inlinable implement take `Substring`. (Which is almost always better in every respect than using `StringProtocol`)

Comment (tbkka) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: Do you mean like this?
```
  @inlinable
  public init?(_ text: String) {
    self.init(substring: text[...])
  }

  public init?(substring: Substring) {
    ... real work ...
  }
```

Comment (jckarter) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: Yeah.

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

File: stdlib/public/core/FloatingPointParsing.swift.gyb (hunk 2 above)
Comment (tbkka) [reply stdlib/public/core/FloatingPointParsing.swift.gyb]: Done.

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
Whew!  I think I _finally_ worked out how to correctly handle the backwards- and forwards- deployment issues here.

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
The CI failed, but looking through the log I don't understand why.  I see no build or test failures.  I do see one inscrutable message:
```
10:18:37 E[10:18:37.417] Transport error: Input/output error
```

**Comment by jckarter** [comment]
@tbkka You might just try again, I'd judge that to be a transient failure.

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
Now the buildbot is failing with a missing symbol `_$sSdySdSgSscfC` -- I'm not yet very good at demangling in my head 😉 but I think this is the new Double constructor I added.  Not sure why it would only fail on this one test, but it seems that I'm not using the right availability declarations?

```
18:42:46 Test Case '-[SourceKitTests.DocumentColorTests testSimple]' started.
18:42:46 dyld: lazy symbol binding failed: Symbol not found: _$sSdySdSgSscfC
18:42:46   Referenced from: /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/sourcekitlsp-macosx-x86_64/x86_64-apple-macosx/release/SourceKitLSPPackageTests.xctest/Contents/MacOS/SourceKitLSPPackageTests
18:42:46   Expected in: /usr/lib/swift/libswiftCore.dylib
```

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
Hmmm....  There's something wrong with how sourcekit-lsp is being tested here:  Some part of that test suite is being built against the new standard library (hence is referencing the new symbol) but is then getting run with the system-installed standard library (the reference to /usr/lib/swift above).

**Comment by benlangmuir** [comment]
> Hmmm.... There's something wrong with how sourcekit-lsp is being tested here: Some part of that test suite is being built against the new standard library (hence is referencing the new symbol) but is then getting run with the system-installed standard library (the reference to /usr/lib/swift above).

This sounds expected to me. We cannot use libswiftCore.dylib from the toolchain on macOS, we have to use the system one at runtime.  Otherwise we couldn't use anything else from the system (e.g. Foundation) without creating a conflict.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
I haven't been able to reproduce this particular test failure locally, so I'll need to keep poking swift-ci to run it for me.

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Comment by shahmishal** [comment]
@swift-ci Please smoke test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
CI tests are all succeeding now, so I went ahead and merged.

**Pull request merged**